sasl/client.go

```diff
@@ -0,0 +1,298 @@
+
+type saslInfoCallback func([]CallbackRequestItem) error
+
+type context struct {
+	State int
+	challenge
+	KIReceive []byte
+	KISend    []byte
+	EncKey    []byte
+	DecKey    []byte
+	HA1       []byte
+}
+
+type client struct {
+	LocalAddress  net.Addr
+	RemoteAddress net.Addr
+	Callback      saslInfoCallback
+	context       *context
+}
+
+type challenge struct {
+	Service    string
+	Hostname   string
+	DigestURI  string
+	Username   string
+	AuthzID    string
+	NonceHex   string
+	CNonceHex  string
+	NC         uint32
+	Algorithm  string
+	QOP        string
+	Realm      string
+	Charset    string
+	MaxBuf     uint32
+	Response   string
+	Method     string
+	SessionKey string
+	Ciphers    []string
+	Cipher     string
+	CipherType cipherType
+}
+
+// Client represents the client portion of the SASL protocol
+type Client interface {
+	FindMech(mechList []string) (mech string, err error)
+	ApplyChallenge(challenge string)
+	MakeResponse() string
+	SetCNonce(string)
+}
+
+// NewClient creates a new instance of the client
+func NewClient(service, hostname string, auth AuthInfo) Client {
+	nonce := make([]byte, saslMD5NoneSize)
+	rand.Read(nonce)
+	return &client{
+		Callback: auth.Callback,
+		context: &context{
+			State: 1,
+			challenge: challenge{
+				Service:   service,
+				Hostname:  hostname,
+				Realm:     "",
+				NC:        0,
+				CNonceHex: base64.StdEncoding.EncodeToString(nonce),
+				MaxBuf:    saslDataMax,
+				Charset:   "utf-8",
+				Method:    "AUTHENTICATE",
+			},
+		},
+	}
+}
+
+func (client *client) FindMech(mechList []string) (mech string, err error) {
+	for _, entry := range mechList {
+		// Eventually this would be the place to switch to different
+		// mech variations, but for now we only support md5
+		if entry == "DIGEST-MD5" {
+			return entry, nil
+		}
+	}
+	return "", fmt.Errorf("SASL: Requested mech not supported")
+}
+
+func md5Sum(a, b []byte) []byte {
+	h := md5.New()
+	h.Write(a)
+	return h.Sum(b)
+}
+
+func (client *client) ApplyChallenge(challenge string) {
+	parseChallenge(challenge, &client.context.challenge)
+}
+
+func minUint32(x, y uint32) uint32 {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+func (client *client) calcSecretHash(password string) []byte {
+	h := md5.New()
+	h.Write([]byte(client.context.Username))
+	h.Write([]byte(":"))
+	h.Write([]byte(client.context.Realm))
+	h.Write([]byte(":"))
+	h.Write([]byte(password))
+	x := h.Sum(nil)
+	return x[:]
+}
+
+func (client *client) calcHA1(password string) []byte {
+	ha1 := md5.New()
+	tmp := client.calcSecretHash(password)
+	if client.context.Algorithm == "md5-sess" {
+		ha1.Write(tmp)
+		ha1.Write([]byte(":"))
+		ha1.Write([]byte(client.context.NonceHex))
+		ha1.Write([]byte(":"))
+		ha1.Write([]byte(client.context.CNonceHex))
+		if client.context.AuthzID != "" {
+			ha1.Write([]byte(":"))
+			ha1.Write([]byte(client.context.AuthzID))
+		}
+	}
+	client.context.HA1 = ha1.Sum(nil)
+	client.context.SessionKey = fmt.Sprintf("%x", client.context.HA1)
```

Comment:
How about `hex.EncodeToString`?  Not a big fan of `fmt.Sprintf` when there's a stdlib function available.